### PR TITLE
#179 ADD HLA module

### PR DESCRIPTION
**Pull request by jalew188**
Move peptdeep_hla module into peptdeep/hla module

### CODE REVIEW

**Comment by jalew188** [comment]
@wahlema peptdeep with HLA will be release in version >=1.3.0

File: requirements/requirements_hla.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+pydivsufsort
```
[the remark this answers is not in the document]
Comment (jalew188) [reply requirements/requirements_hla.txt:1]: Yes, Maria will work on this and the tutorial.

File: peptdeep/hla/hla_class1.py [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+        self.predict_df = precursor_df
+
+    def _prepare_train_data_df(self,
+        precursor_df, **kwargs
+    ):
+        self.__training = True
+        precursor_df['nAA'] = precursor_df.sequence.str.len()
+        precursor_df.drop(
+            index=precursor_df[
+                (precursor_df.nAA<8)|(precursor_df.nAA>14)
```
[the remark this answers is not in the document]
Comment (jalew188) [reply peptdeep/hla/hla_class1.py]: Done

File: peptdeep/hla/hla_class1.py [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+            self.min_peptide_length,
+            self.max_peptide_length
+        )
+
+    def _predict_all_probs(self, digest_batch_size):
+        for i in tqdm.tqdm(range(
+            0, len(self.digested_idxes_df),
+            digest_batch_size
+        )):
+            _df = self.digested_idxes_df.iloc[i:i+digest_batch_size]
```
[the remark this answers is not in the document]
Comment (jalew188) [reply peptdeep/hla/hla_class1.py:252]: I used _xxx for temp variables.

File: peptdeep/hla/hla_class1.py [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+            Peptides above this probability are kept, by default 0.7
+        digest_batch_size : int, optional
+            Batch size for digestion, by default 1024000
+
+        Returns
+        -------
+        pd.DataFrame
+            The peptide dataframe in alphabase format.
+        """
+        if isinstance(protein_data, pd.DataFrame):
```
[the remark this answers is not in the document]
Comment (jalew188) [reply peptdeep/hla/hla_class1.py:309]: Can be

File: peptdeep/hla/hla_class1.py [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+        self.min_peptide_length = min_peptide_length
+        self.max_peptide_length = max_peptide_length
+        self._n_neg_per_pos_training = 1
+
+        self.predict_batch_size = 4096
+
+    def _prepare_predict_data_df(self,
+        precursor_df:pd.DataFrame,
+    ):
+        self.__training = False
```
[the remark this answers is not in the document]
Comment (jalew188) [reply peptdeep/hla/hla_class1.py:181]: Done

File: peptdeep/hla/hla_class1.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jalew188) [reply peptdeep/hla/hla_class1.py]: Done

File: peptdeep/pretrained_models.py [hunk 6]
```diff
@@ -83,6 +85,10 @@ def download_models(
         Remote or local path.
         Defaults to `peptdeep.pretrained_models.model_url`
 
+    target_path : str, optional
+        Target file path after download.
+        Defaults to `peptdeep.pretrained_models.model_zip`
```
[the remark this answers is not in the document]
Comment (jalew188) [reply peptdeep/pretrained_models.py:90]: This a variable.

File: peptdeep/hla/hla_utils.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    -------
+    pd.DataFrame
+        protein dataframe
+    """
+    if isinstance(protein_data, str):
+        protein_dict = load_all_proteins([protein_data])
+    elif isinstance(protein_data, (list,tuple,set)):
+        protein_dict = load_all_proteins(protein_data)
+    elif isinstance(protein_data, str):
+        protein_dict = load_all_proteins([protein_data])
```
[the remark this answers is not in the document]
Comment (jalew188) [reply peptdeep/hla/hla_utils.py:36]: Fixed

File: peptdeep/hla/hla_utils.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    elif isinstance(protein_data, dict):
+        protein_dict = protein_data
+    else:
+        return pd.DataFrame()
```
[the remark this answers is not in the document]
Comment (jalew188) [reply peptdeep/hla/hla_utils.py:40]: Add TypeError here

File: peptdeep/hla/hla_class1.py [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+)
+
+
+_model_zip_name = global_settings['local_hla_model_zip_name']
+_model_url = global_settings['hla_model_url']
+_model_zip = os.path.join(
+    pretrain_dir, _model_zip_name
+)
+
+if not os.path.exists(_model_zip):
```
[the remark this answers is not in the document]
Comment (jalew188) [reply peptdeep/hla/hla_class1.py:30]: Yes, done

**Review by wahlema — APPROVED** [review]
Done :)

File: peptdeep/hla/hla_utils.py [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+    )
+
+
+@numba.njit
+def check_sty(seq):
+    for aa in seq:
+        if aa in "STY":
+            return True
+    return False
+
```
[the remark this answers is not in the document]
Comment (jalew188) [reply peptdeep/hla/hla_utils.py:136]: This function was designed for your HLA-phos-peptide prediction, but it is not used anyway.

[the remark this answers is not in the document]
Comment (jalew188) [reply peptdeep/hla/hla_utils.py:136]: I will change it to private func

File: peptdeep/hla/hla_class1.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mschwoer) [reply peptdeep/hla/hla_class1.py]: well, in a strict sense, any variable that is not returned is temporary, right?  :-)

[the remark this answers is not in the document]
Comment (jalew188) [reply peptdeep/hla/hla_class1.py]: Yes, but there is no rules for temp variable names

[the remark this answers is not in the document]
Comment (jalew188) [reply peptdeep/hla/hla_class1.py]: Sometimes I used this kind of rule for a huge obj:
```
for _df in df_list:
    _df is only used inside the for loop, and will not be used outside
```

**Pull request merged**